### ISSUE TITLE
Narrator is not reading the state(Checked/Unchecked) of the DetailsList selection checkbox

### ISSUE THREAD

**Issue by utkarshagoswami** [OfficeDev/office-ui-fabric-react#6525]
### Bug Report 
 
 
- __Package version(s)__: (fill this out) 
- __Browser and OS versions__: Microsoft Edge: 42.17134.1.0

#### Priorities and help requested:
 
Are you willing to submit a PR to fix? (Yes, No)
 
Requested priority: (Blocking, High, Normal, Low) 
 
Products/sites affected: (if applicable)
 
#### Describe the issue:

On taking the control to any row involving checkbox, the spacebar can check/uncheck the item. The NVDA/Narrator does not announce that. Visually Impaired Users may find difficulty in using the application if the screen reader does not read the contents and controls properly
 
#### Actual behavior:
 
Narrator is not reading the state(Checked/Unchecked) of the checkbox.
 
#### Expected behavior:

Narrator should read the state(Checked/Unchecked) of the checkbox when space bar is pressed.

**Comment by Vitalius1** [comment]
@utkarshagoswami I did some tests in here https://developer.microsoft.com/en-us/fabric#/components/checkbox with macOS voiceover and it announces for me the state of checked/unchecked.
Do you have a codepen repro where in your case with NVDA or Narrator it's not announcing?
Use this a starter https://codepen.io/dzearing/pen/eKvjKN. Can you fork it and modify to test it on your end? This would help us to exclude the possibility if it's a screenreader issue. Thanks

**Comment by utkarshagoswami** [comment]
The issue is with the DetailsList. There is no way to reach the checkbox in a row via keyboard and the only way to select or unselect a row is by selecting the complete row and then using Space Bar. When doing this, Narrator and NVDA are both announcing "Space" instead of the state checked/unchecked. If I use mouse to reach the checkbox, the announcement done is in lieu with the state. The issue to be solved is for the keyboard to be able to reach checkbox on each row or for the state to be announced when Space Bar is used on the whole row

**Comment by KevinTCoughlin** [comment]
I can confirm that Narrator is not reading the selected state of the row even though we are toggling `aria-selected` per row on `SPACE`.

```
Space
Item 3 3, non-selected, 
Space
Space
Space
Item 4 4, selected, 
Space
Space
Space
Space
Space
Space
```

`aria-selected` seems like the correct choice given that the `role="row"` is set. The interaction pattern the component currently has doesn't lend itself to make the checkbox independently focusable.

Next, I'll see if Narrator is not announcing the change in `aria-selected` state because the row is focused and therefore already implicitly selected as mentioned https://github.com/nvaccess/nvda/issues/4816#issuecomment-155333266. Perhaps there is a workaround to get `aria-selected` to better report its state to Narrator.


**References:**

- https://www.w3.org/TR/wai-aria-1.1/#aria-selected

**Comment by KevinTCoughlin** [comment]
Also related #6469

**Comment by cliffkoh** [comment]
It is actually possible to reach the checkbox via scan mode, and in this case the toggle of the state does read the change in state, in Narrator.

**Comment by cliffkoh** [comment]
@KevinTCoughlin Actually, I think we want `aria-checked` instead. Putting `aria-checked` on the row (in addition to the checkbox) seems like it would work and would be reasonable given that when not in scan mode, keyboarding would have you move up and down the rows.

**Comment by KevinTCoughlin** [comment]
> 
> 
> @KevinTCoughlin Actually, I think we want `aria-checked` instead. Putting `aria-checked` on the row (in addition to the checkbox) seems like it would work and would be reasonable given that when not in scan mode, keyboarding would have you move up and down the rows.

Regarding this approach, I'm unsure that `aria-checked` will be valid given the row's `role="row"`.

Characteristic | Value
-- | --
Used in Roles | checkbox option radio switch
Inherits into Roles: | menuitemcheckbox menuitemradio treeitem
Value: | tristate

https://www.w3.org/TR/wai-aria-1.1/#aria-checked

**Comment by KevinTCoughlin** [comment]
_Still_ debugging the root cause. I noticed two issues while debugging this today:

1. The Narrator read-out can fall out-of-sync with the UI re: `aria-selected` state. I'm still not sure what the exact steps are. Howeverful, a few times I was able to multi-select, then unselect individual items, and Narrator would read-out the inverse of their correct state.
2. Narrator reads out `aria-selected` state for `click`, `UP` & `DOWN`, but not `SPACE`. Unsure why `SPACE` in particular.

Re: the first issue I mentioned in this comment, it almost seems like the Narrator's cache of elements is not working correctly with how React is replacing the row elements. Depending on selection state, react renders either a selected or unselected row. Each has their own unique id.

**Comment by JasonGore** [comment]
I just did some analysis as part of #6478 with regards to `DetailsList` role hierarchy and how `row` must be owned by a particular element that `DetailsList` appears to violate. I wonder if this role violation could be causing confusion with screen readers.

**Comment by KevinTCoughlin** [comment]
> 
> 
> I just did some [analysis](https://github.com/OfficeDev/office-ui-fabric-react/issues/6478#issuecomment-435527323) as part of #6478 with regards to `DetailsList` role hierarchy and how `row` must be owned by a particular element that `DetailsList` appears to violate. I wonder if this role violation could be causing confusion with screen readers.

I think you're onto something. In my testing earlier this week, if I swapped the `role="list"` on the inner `List` sections used for pagination with something like `grid` it seemed to announce correctly. I was using `tablist` and `tab` for debug since the w3 example for `aria-selected` used them.

**Comment by JasonGore** [comment]
I mixed up `DetailsList` with `GroupedList` in my comment but I feel like `DetailsList` could have similar issues. I'll edit my comment to correct.

**Comment by ecraig12345** [comment]
@sonimishra just filed this same issue under #7436 (that repro is specific to PowerApps but the root issue is the same). Do we know when this will be addressed?

**Comment by KevinTCoughlin** [comment]
> 
> 
> @sonimishra just filed this same issue under #7436 (that repro is specific to PowerApps but the root issue is the same). Do we know when this will be addressed?

We have some a11y contacts that may help us debug this shortly after the holidays at maximum (I think). That's assuming that this is solvable (I hope so).

I'll sync with @natalieethell and take another pass and report my findings. I imagine the bug mainly has to do with following the [role="listbox"](https://www.w3.org/TR/2017/WD-wai-aria-practices-1.1-20170628/examples/listbox/listbox.html) spec so that the Accessibility Tree is correct.

Its a pretty egregious accessibility bug, but not the most straight-forward to debug.

**Comment by sonimishra** [comment]
+ Prateek

**Comment by anubhavaggarwal22** [comment]
@KevinTCoughlin  we have a blocker accessibility bug (https://msazure.visualstudio.com/One/_queries/edit/3571922/?triage=true) on flow team because of this  bug, is there a ETA when this will be fixed?

**Comment by KevinTCoughlin** [comment]
> @KevinTCoughlin we have a blocker accessibility bug on flow team because of this bug, is there a ETA when this will be fixed?

@anubhavaggarwal22, I don't recommend linking to internal boards on a public forum. Let's start an internal thread to discuss. I am not an expert in this area and have been unable to solve myself.

Have you considered a work around such as using [aria-live](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) region(s) to announce the selection state while we arrive on a solution (if one exists)?

Adding @cschlechty for any insight to what is stopping `aria-selected` state from being read when used in the `DetailsList`.

**Comment by bmathwig** [comment]
@KevinTCoughlin Has there been any updates on this issue? It looks like the DevOps task for the Flow team was marked "Done" via exclusion. Although we aren't using Fabric React components directly, I'm hitting this exact same issue with Narrator and non-Checkbox role elements. Curious if this is a Narrator tree refresh bug or if you were able to find a work around.

**Comment by xugao** [comment]
I hear NVDA is announcing the info correctly. try [this pen](https://codepen.io/xugao/pen/bGNyEMw).
I will file a bug to the Narrator team.
Meanwhile, @bmathwig @utkarshagoswami , you can use `<Announced />` component to leverage `aria-live` to walk around this issue

**Comment by cschlechty** [comment]
Please don't hack it. Hacks tend to propagate and create more work later.
Lean on narrator.

**Comment by xugao** [comment]
this is fixed by narrator team [here](https://microsoft.visualstudio.com/OS/_workitems/edit/24793756)
